Add `gulp-imagemin` task

claire-codes opened the pull request:
When the task `gulp imagemin` is run, it finds all the images in the `web/images` directory within each theme in `app/design` and minifies them if possible.

* The files are minified in place
* There is a default config provided which can be overwritten by create a `imageMinConfig.json` file in the normal place in `dev/tools`.
* Package used is [https://github.com/sindresorhus/gulp-imagemin](https://github.com/sindresorhus/gulp-imagemin)

I would have liked to add a caching task that would only attempt to minify images that are new or updated but I couldn't work out how to get it to work with frontools helper, tasks and plugins ecosystem.

I would also like to be able to print out a message after the images have finished minifying per theme.

claire-codes committed:
Add gulp-imagemin task

claire-codes committed:
Add updated yarn.lock

Still in the habit of `npm i`

claire-codes committed:
Put imageMin config into external file

Adds a try-catch so that if the user doesn't have a config file set up
then they will get some default config options.

claire-codes committed:
Add new task to readme

jakestubbz commented on helper/imagemin.js:
```diff
@@ -0,0 +1,25 @@
+        // Checks for project specific config in dev/tools/frontools/config
+        imageMinConfig = require('../helper/config-loader')('imageMinConfig.json', plugins, config);
+    } catch (e) {
+        imageMinConfig = { "optimizationLevel": 3, "progressive": true, "interlaced": true };
```
We don't need to do this try/catch, if you check in the `config-loader` you'll see it throws an error if the config isn't found.

We should create a .sample config in the `config/` directory with the recommended defaults, it'll get copied over to the project on `gulp setup` and if it's then manually removed the `config-loader` module will throw out the error.

claire-codes replied:
Aaaah I didn't realise how that it is how it works 👍

jakestubbz commented on helper/imagemin.js:
```diff
@@ -0,0 +1,25 @@
+
+    try {
+        // Checks for project specific config in dev/tools/frontools/config
+        imageMinConfig = require('../helper/config-loader')('imageMinConfig.json', plugins, config);
```
We can go with `imagemin.json` to keep to the current naming convention used by the other config files.

jakestubbz commented on helper/imagemin.js:
```diff
@@ -0,0 +1,25 @@
+
+          // Look for any images in the web/images directory and minify them in place
+          imageDir       = srcBase + '/web/images/**/*',
+          minnedImageDir = srcBase + '/web/images/';
```
Do you think it's worth running this on `/web/icons` too? I'm not sure how/if that'll impact the icon task... I seem to recall that gulpicon can only optimise PNGs.

claire-codes replied:
As to running it on different directories, I was thinking that we could run it on this directory by default and then optionally check for an argument to specify a different folder?

I've tested this on an SVG and it did manage to minify it from 4kb to 2kb! And the repo says "Minify PNG, JPEG, GIF and SVG images with imagemin"

jakestubbz commented on helper/imagemin.js:
```diff
@@ -0,0 +1,25 @@
+
+    return gulp.src(imageDir)
+        .pipe(plugins.imagemin(imageMinConfig))
+        .pipe(gulp.dest(minnedImageDir));
```
If you want to log a message once the task is finished, you can pipe to the logger here (see L20 `helpers/css-lint.js` for an example).

jakestubbz commented:
>I would have liked to add a caching task

@claireparker: Have you got a plugin in mind that you'd like to use?

claireparker commented:
@jakestubbz I wanted to use plain old `gulp-cache` as in this example https://markgoodyear.com/2014/01/getting-started-with-gulp/ (go to the "Compress Images" paragraph) which I've managed to set up on projects with less complex gulp files 😑 but couldn't get it to work properly with this setup.

claire-codes committed:
Remove try-catch and add sample config data

Don't need try-catch to check for config file and provide a fallback -
this is not how config-helper is designed to work. Instead we should
prvide a sample file in the config directory which will be copied to the
local project on `gulp setup`. Then if this doesn't exist, the gulp task
should correctly fail.

claire-codes committed:
Add logging after task has finished per file

claire-codes committed:
Add a comma to make logging message read better

jakestubbz commented:
@claireparker: Okay, cool. I'll have a read and see if I can figure out how we can plug it into frontools - it'd be useful to not have it running unless required.

claireparker committed:
Merge branch 'master' of https://github.com/AmpersandHQ/magento2-frontools into HEAD

Also run `yarn` to make sure `yarn.lock` is up to date correctly

claireparker committed:
Add --dir flag to imagemin task

claireparker commented:
@jakestubbz I've updated to allow a flag which provides a directory of the user's choice to minify. This means that you can do

```
gulp imagemin --dir dev/tests/backstop_data/bitmaps_reference/
```

to minify all the Backstop screenshots!

Still outstanding is adding a caching task here - currently it tries to minify everything each time the task is run.
